File: modules/coding/data_structures/object-oriented-programming/c++/oop.cpp
// https://www.tutorialspoint.com/cplusplus/cpp_classes_objects.htm
// g++ -o oop ./oop.cpp && ./oop


#include <iostream>

using namespace std;

class Box {

   public: //public members

      int width;
      int height;
      char color[10];
      int version;

      // member functions

      int get_area() {

         return calculate_area();
      }

      int get_perimeter() {

         return calculate_perimeter();
      }

      int get_version() {

         return version;
      }

      static void print_static_text() {

        cout << "print static text";
      }

      char * print_box() {

        sprintf(print_output, "(w:%d,h:%d)", width, height);

        // cout << "(w:" << width << ",h:" << height << ")\n" << endl;
        return print_output;
      }

      // constructor declaration - defined in Box::Box() below.
      Box(int width, int height);

      // destructor declaration - defined in Box::~Box() below.
      ~Box();

    protected: // protected members

      char print_output[50];

      int calculate_area() {

         return (width * height);
      }

    private:

      int calculate_perimeter() {

         return (width + height) * 2;
      }

};

// constructor function
Box::Box(int width, int height) {

  this->width = width;

  this->height = height;

  sprintf(this->color, "Red");
}

// destructor function
Box::~Box(void) {

   cout << "Box is being deleted" << endl;

   cout << this->print_box() << endl;
}

class Cube: public Box {

   public:

     int depth;

     Cube(int width, int height, int depth);

     int get_volume() {

       int volume = (width * height * depth);

        return volume;
     }

     int get_area() {

       int volume = (width * height * depth);

        return volume;
     }

     char * print_cube() {

       sprintf(print_output, "(w:%d,h:%d,d:%d)", width, height, depth);

       // cout << "(w:" << width << ",h:" << height << ")\n" << endl;
       return print_output;
     }
};

Cube::Cube(int width, int height, int depth) : Box(width, height) {

  this->width = width;

  this->height = height;

  this->depth = depth;

  sprintf(this->color, "Blue");
}

int main() {

   Box box(10, 20);

   cout << "New box : " << box.print_box() << endl;

   box.version = 1;

   cout << "Width of the box is : " << box.width << endl;
   cout << "Height of the box is : " << box.height << endl;
   cout << "Color of the box is : " << box.color << endl;
   cout << "Area of the box is : " << box.get_area() << endl;
   cout << "Perimeter of the box is : " << box.get_perimeter() << endl;
   cout << "Version of the box is : " << box.get_version() << endl;
   cout << "Static text of the box is : "; box.print_static_text(); cout << endl;

   Cube cube(10, 20, 30);

   cout << "\nNew cube : " << cube.print_cube() << endl;

   cout << "Width of the cube is : " << cube.width << endl;
   cout << "Height of the cube is : " << cube.height << endl;
   cout << "Depth of the cube is : " << cube.height << endl;
   cout << "Color of the cube is : " << cube.color << endl;
   cout << "Area of the cube is : " << cube.get_area() << endl;
   cout << "Perimeter of the cube is : " << cube.get_perimeter() << endl;
   cout << "Volume of the cube is : " << cube.get_volume() << endl;
 }

File: modules/coding/data_structures/object-oriented-programming/ruby/oop.rb
# define a class
class Box

  attr_accessor :object_color

  # class variable
  @@version = 0

  # constructor method
   def initialize(width, height)

      @object_width = width
      @object_height = height
      @object_color = "Red"
   end

   # to_s method
   def to_s
      "(w:#@object_width,h:#@object_height)"  # string formatting of the object.
   end

   # accessor methods
   # ----------------

   def get_width()

      @object_width
   end

   def get_height()

      @object_height
   end

   def get_color()

      @object_color
   end

   # setter methods
   # --------------

   def set_width(new_width)

      @object_width = new_width
   end

   def set_height(new_height)

      @object_height = new_height
   end

   def set_color(new_color)

      @object_color = new_color
   end

   # instance methods
   # ----------------

   def calculate_perimeter(width, height)

     (@object_width + @object_height) * 2
   end

   def calculate_area(width, height)

     @object_width * @object_height
   end

   # set private methods
   private :calculate_area

   def get_area()

      calculate_area(@object_width, @object_height)
   end

   def get_perimeter()

     self.calculate_perimeter(@object_width, @object_height)
   end

   # set protected methods
   protected :calculate_perimeter

   def print_version_instance_method()

      @@version
   end

   # class methods

   def self.print_version_class_method()

      @@version
   end

   def self.increment_version_class_method()

      @@version += 1
   end
end

# create an object
box = Box.new(10, 20)

puts "New box: #{box}"

# update box color via attr_accessor
box.object_color = "Blue"

# use accessor methods
width = box.get_width()
height = box.get_height()
color = box.get_color()
area = box.get_area()
perimeter = box.get_perimeter()

puts "Width of the box is : #{width}"
puts "Height of the box is : #{height}"
puts "Color of the box is : #{color}"
puts "Area of the box is : #{area}"
puts "Perimeter of the box is : #{perimeter}"

Box.increment_version_class_method()

puts "Incremented version of the box is (Instance Method): #{box.print_version_instance_method()}"
puts "Incremented version of the box is (Class Method): #{Box.print_version_class_method()}"

class Cube < Box

  def initialize(width, height, depth)

     @object_width = width
     @object_height = height
     @object_depth = depth
  end

  # to_s method
  def to_s

     "(w:#@object_width,h:#@object_height,d:#@object_depth)"  # string formatting of the object.
  end

  def get_depth()

    @object_depth
  end

  def get_volume()

    volume =  @object_width * @object_height * @object_depth

    volume
  end

  def get_area()

     super + 9000
  end
end


# create an object
cube = Cube.new(10, 20, 30)

puts "\nNew cube: #{cube}"

# use accessor methods
width = cube.get_width()
height = cube.get_height()
depth = cube.get_depth()
area = cube.get_area()
perimeter = cube.get_perimeter()
volume = cube.get_volume()

puts "Width of the cube is : #{width}"
puts "Height of the cube is : #{height}"
puts "Depth of the cube is : #{depth}"
puts "Area of the cube is : #{area}"
puts "Perimeter of the cube is : #{perimeter}"
puts "Volume of the cube is : #{volume}"
